cli/updater/updater.go

```diff
@@ -0,0 +1,138 @@
+// This file is part of arduino-cli.
+//
+// Copyright 2020 ARDUINO SA (http://www.arduino.cc/)
+//
+// This software is released under the GNU General Public License version 3,
+// which covers the main part of arduino-cli.
+// The terms of this license can be found at:
+// https://www.gnu.org/licenses/gpl-3.0.en.html
+//
+// You can be released from the requirements of the above licenses by purchasing
+// a commercial license. Buying such a license is mandatory if you want to
+// modify or otherwise use the software for commercial activities involving the
+// Arduino software without disclosing the source code of your own applications.
+// To purchase a commercial license, send an email to license@arduino.cc.
+
+package updater
+
+import (
+	"os"
+	"strings"
+	"time"
+
+	"github.com/arduino/arduino-cli/cli/feedback"
+	"github.com/arduino/arduino-cli/cli/globals"
+	"github.com/arduino/arduino-cli/configuration"
+	"github.com/arduino/arduino-cli/httpclient"
+	"github.com/arduino/arduino-cli/i18n"
+	"github.com/arduino/arduino-cli/inventory"
+	"github.com/fatih/color"
+	semver "go.bug.st/relaxed-semver"
+)
+
+var tr = i18n.Tr
+
+// CheckForUpdate returns the latest available version if greater than
+// the one running and it makes sense to check for an update, nil in all other cases
+func CheckForUpdate(currentVersion *semver.Version) *semver.Version {
+	if !shouldCheckForUpdate(currentVersion) {
+		return nil
+	}
+
+	return checkForUpdate(currentVersion)
+}
+
+// ForceCheckForUpdate always returns the latest available version if greater than
+// the one running, nil in all other cases
+func ForceCheckForUpdate(currentVersion *semver.Version) *semver.Version {
+	return checkForUpdate(currentVersion)
+}
+
+func checkForUpdate(currentVersion *semver.Version) *semver.Version {
+	defer func() {
+		// Always save the last time we checked for updates at the end
+		inventory.Store.Set("updater.last_check_time", time.Now())
+		inventory.WriteStore()
+	}()
+
+	latestVersion, err := semver.Parse(getLatestRelease())
+	if err != nil {
+		return nil
+	}
+
+	if currentVersion.GreaterThanOrEqual(latestVersion) {
+		// Current version is already good enough
+		return nil
+	}
+
+	return latestVersion
+}
+
+// NotifyNewVersionIsAvailable prints information about the new latestVersion
+func NotifyNewVersionIsAvailable(latestVersion string) {
+	feedback.Errorf("\n\n%s %s → %s\n%s",
+		color.YellowString(tr("A new release of Arduino CLI is available:")),
+		color.CyanString(globals.VersionInfo.VersionString),
+		color.CyanString(latestVersion),
+		color.YellowString("https://arduino.github.io/arduino-cli/latest/installation/#latest-packages"))
```

Comment:
Is this URL going to stay? Will it move at some point from `arduino.github.io` to `arduino.cc/something`?

Reply:
Ideally we'll change it in the future when the Arduino CLI will be added to https://www.arduino.cc/en/software.